src/rewindable_range_proof.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    commitment::HomomorphicCommitment,
+    errors::RangeProofError,
+    keys::{PublicKey, SecretKey},
+};
+
+pub const REWIND_USER_MESSAGE_LENGTH: usize = 21;
+
+pub trait RewindableRangeProofService {
+    type P: Sized;
+    type K: SecretKey;
+    type PK: PublicKey<K = Self::K>;
+
+    /// Construct a rangeproof with the ability to rewind it. Requires two rewind keys and a 19-byte message to be
+    /// included in the range proof. The proof can contain 23 bytes but 4 bytes are used to confirm that a rewind
+    /// was performed correctly
+    fn construct_proof_with_rewind_key(
+        &self,
+        key: &Self::K,
+        value: u64,
+        rewind_key: &Self::K,
+        rewind_blinding_key: &Self::K,
+        proof_message: &[u8; REWIND_USER_MESSAGE_LENGTH],
```

Comment:
Fine for this PR, since it was like this, but ideally, this should be general and not locked to 21 bytes for every impl of the trait. `&[u8]` will work here.